### PR TITLE
chore: remove bugsnag

### PR DESCRIPTION
**Pull request by lvrach**
# Description

Remove bugsnag from rudder-server, capturing panics using logs instead.

## Linear Ticket

https://linear.app/rudderstack/issue/PIPE-1435/remove-bug-snag

Resolves PIPE-1435
Resolves PIPE-611

## Security

- [x] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

File: utils/crash/noop.go [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package crash
+
+type NOOP struct{}
+
+func (n *NOOP) Notify(team string) func() {
```
Comment (fracasula) [inline utils/crash/noop.go:5]:
```suggestion
func (*NOOP) Notify(_ string) func() {
```

File: utils/crash/logger.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+				p.logger.Fatalw("Panic detected. Application will crash.",
+					"stack", string(debug.Stack()),
+					"panic", r,
+					"team", team,
+					"goRoutines", runtime.NumGoroutine(),
+					"version", p.opts.AppVersion,
+					"releaseStage", p.opts.ReleaseStage,
+					"appType", p.opts.AppType,
+				)
+				logger.Sync()
```
Comment (fracasula) [inline utils/crash/logger.go:38]: Doesn't this do a `Default.Sync()`? Would it affect correctly `p.logger` also? 
We also have a `Sync` method in the logger factory, I didn't reverse engineered the whole thing though.

File: utils/crash/logger.go [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+				panic(r)
+			})
+		}
+	}
+}
```
Comment (achettyiitr) [inline utils/crash/logger.go:43]: Does it it needs to be exported?
```suggestion
type panicLogger struct {
	notifyOnce sync.Once
	logger     logger.Logger

	opts PanicWrapperOpts
}

func usingLogger(logger logger.Logger, opts PanicWrapperOpts) panicHandler {
	return &panicLogger{
		logger: logger.Child("panic"),
		opts:   opts,
	}
}

func (p *panicLogger) Notify(team string) func() {
	return func() {
		if r := recover(); r != nil {
			p.notifyOnce.Do(func() {
				p.logger.Fatalw("Panic detected. Application will crash.",
					"stack", string(debug.Stack()),
					"panic", r,
					"team", team,
					"goRoutines", runtime.NumGoroutine(),
					"version", p.opts.AppVersion,
					"releaseStage", p.opts.ReleaseStage,
					"appType", p.opts.AppType,
				)
				logger.Sync()
				panic(r)
			})
		}
	}
}
```

File: utils/crash/logger.go [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+			p.notifyOnce.Do(func() {
+				p.logger.Fatalw("Panic detected. Application will crash.",
+					"stack", string(debug.Stack()),
+					"panic", r,
+					"team", team,
+					"goRoutines", runtime.NumGoroutine(),
+					"version", p.opts.AppVersion,
+					"releaseStage", p.opts.ReleaseStage,
+					"appType", p.opts.AppType,
+				)
```
Comment (achettyiitr) [inline utils/crash/logger.go:41]: We can use a non-sugared structured log in here.
```suggestion
				p.logger.Fataln("Panic detected. Application will crash.",
					logger.NewStringField("stack", string(debug.Stack())),
					logger.NewField("panic", r),
					logger.NewStringField("team", team),
					logger.NewIntField("goRoutines", int64(runtime.NumGoroutine())),
					logger.NewStringField("version", p.opts.AppVersion),
					logger.NewStringField("releaseStage", p.opts.ReleaseStage),
					logger.NewStringField("appType", p.opts.AppType),
				)
```

File: utils/crash/default.go [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+	AppVersion   string
+	ReleaseStage string
+	AppType      string
+}
+
+func Configure(logger logger.Logger, opts PanicWrapperOpts) {
+	Default = UsingLogger(logger, opts)
+}
+
+func WrapperForWarehouse(fn func() error) func() error {
```
Comment (fracasula) [inline utils/crash/default.go:26]: Personal preference probably but I'd call this `NotifyWarehouse` instead.

File: rruntime/goroutine-factory.go [hunk 6]
```diff
@@ -23,16 +19,14 @@ import (
...
+		defer crash.Notify("Core")()
 		function()
 	}()
 }
 
 func GoForWarehouse(function func()) {
 	go func() {
-		ctx := bugsnag.StartSession(context.Background())
-		defer misc.BugsnagNotify(ctx, "Warehouse")()
+		defer crash.Notify("Warehouse")()
```
Comment (fracasula) [inline rruntime/goroutine-factory.go:29]: Can't we use the wrapper here?

Comment (fracasula) [reply rruntime/goroutine-factory.go:29]: Same for core.

**Review by fracasula — APPROVED** [review]
Approved with minor comments :+1:

File: rruntime/goroutine-factory.go (hunk 6 above)
Comment (lvrach) [reply rruntime/goroutine-factory.go:29]: The wrapper requires a diffrent signature for function, func() error.

**Pull request merged**